hal/common/retarget.cpp

```diff
@@ -94,24 +95,30 @@ FileHandle::~FileHandle() {
 }
 
 #if DEVICE_SERIAL
-extern int stdio_uart_inited;
-extern serial_t stdio_uart;
 #if MBED_CONF_CORE_STDIO_CONVERT_NEWLINES
 static char stdio_in_prev;
 static char stdio_out_prev;
 #endif
-#endif
 
-static void init_serial() {
-#if DEVICE_SERIAL
-    if (stdio_uart_inited) return;
-    serial_init(&stdio_uart, STDIO_UART_TX, STDIO_UART_RX);
+namespace mbed {
+
+Serial& get_stdio_serial()
+{
+    static bool stdio_uart_inited = false;
+    static Serial stdio_serial(STDIO_UART_TX, STDIO_UART_RX);
```

Comment:
> Unfortunately, C++ is very terrible at garbage-collecting globally-allocated non-pod objects. It may not be much of a concern here, since we expect almost all devices to have stdio, but if a device never calls a serial function, this static Serial declaration will cause the entire serial codebase to get dragged into the final image.

I am not certain this is a case. It's not in the local scope. I just tested both solutions, this solution and with placement new, the only difference for IAR was a static guard added for this one (as you noted, one static guards are present). I'll update it to the placement new, thanks for the tip, as it saves some memory (see my commit for more info)